tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c

```diff
@@ -0,0 +1,250 @@
+#include "mesh_test.h"
+#include "mesh/net.h"
+#include "mesh/beacon.h"
+#include "mesh/mesh.h"
+#include "mesh/foundation.h"
+
+#define LOG_MODULE_NAME test_beacon
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#define GROUP_ADDR 0xc000
+#define WAIT_TIME 60 /*seconds*/
+#define PROVISIONED_INTERVAL       K_SECONDS(10)
+
+extern enum bst_result_t bst_result;
+
+static const struct bt_mesh_test_cfg tx_cfg = {
+	.addr = 0x0001,
+	.dev_key = { 0x01 },
+};
+static const struct bt_mesh_test_cfg rx_cfg = {
+	.addr = 0x0002,
+	.dev_key = { 0x02 },
+};
+
+static void test_tx_init(void)
+{
+	bt_mesh_test_cfg_set(&tx_cfg, WAIT_TIME);
+}
+
+static void test_rx_init(void)
+{
+	bt_mesh_test_cfg_set(&rx_cfg, WAIT_TIME);
+}
+
+__attribute__((unused)) static void ivu_log(void)
+{
+	LOG_INF("ivi: %i", bt_mesh.iv_index);
+	LOG_INF("flags:");
+
+	if (atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_INITIATOR)) {
+		LOG_INF("IVU initiator");
+	}
+
+	if (atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_IN_PROGRESS)) {
+		LOG_INF("IVU in progress");
+	}
+
+	if (atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_PENDING)) {
+		LOG_INF("IVU pending");
+	}
+
+	if (atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_TEST)) {
+		LOG_INF("IVU in test mode");
+	}
+}
+
+static void test_tx_on_iv_update(void)
+{
+	bt_mesh_test_setup();
+	ASSERT_OK(atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_INITIATOR));
+	ASSERT_OK(atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_IN_PROGRESS));
+	ASSERT_OK(atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_PENDING));
+	ASSERT_OK(atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_TEST));
+	ASSERT_OK(bt_mesh.iv_index != 0);
+
+	/* shift beaconing time line to avoid boundary cases. */
+	k_sleep(K_SECONDS(1));
+
+	bt_mesh_iv_update_test(true);
+	ASSERT_OK(!atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_TEST));
+
+	ASSERT_OK(!bt_mesh_iv_update());
+	ASSERT_OK(!atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_IN_PROGRESS));
+	ASSERT_OK(bt_mesh.iv_index != 1);
+
+	k_sleep(PROVISIONED_INTERVAL);
+
+	ASSERT_OK(bt_mesh_iv_update());
+	ASSERT_OK(atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_IN_PROGRESS));
+	ASSERT_OK(bt_mesh.iv_index != 1);
+
+	k_sleep(PROVISIONED_INTERVAL);
+
+	ASSERT_OK(!bt_mesh_iv_update());
+	ASSERT_OK(!atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_IN_PROGRESS));
+	ASSERT_OK(bt_mesh.iv_index != 2);
+
+	k_sleep(PROVISIONED_INTERVAL);
+
+	PASS();
+}
+
+static void test_rx_on_iv_update(void)
+{
+	bt_mesh_test_setup();
+	/* disable beaconing from Rx device to prevent
+	 * the time line adaptation due to observation algorithm.
+	 */
+	bt_mesh_beacon_disable();
+	ASSERT_OK(atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_INITIATOR));
+	ASSERT_OK(atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_IN_PROGRESS));
+	ASSERT_OK(atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_PENDING));
+	ASSERT_OK(atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_TEST));
+	ASSERT_OK(bt_mesh.iv_index != 0);
+
+	/* shift beaconing time line to avoid boundary cases. */
+	k_sleep(K_SECONDS(1));
+
+	bt_mesh_iv_update_test(true);
```

Comment:
Hm. I did a 1-second time shift especially to avoid race conditions. The first beacon is sent 1 second before the test mode setting, didn't that?

Reply:
I mean, we need test mode to be active to be able to accept IV updates in the first beacon. Since both devices wait 1 second before they start acting, the RX device will enable test mode at the same time as the TX device starts sending the first beacon. I think the 1 second delay should stay, but that we should enable test mode before it

Reply:
Devices start acting (mesh stack) immediately. Test scenario starts acting with 1 second delay.
That means the first beacon is transmitted immediately. Then scenario waits for 1 second and starts updating test mode, initiates IV update procedure. That doesn't cause immediate beacon emitting (excluding proxy beacons). The stack will recalculate beacon fields and will start transmitting according to the scheduled time(considering observation).  That means the next beacon (that considers IV update or key refresh changes) will be transmitted only after 9 seconds.
Beacon is sent immediately only if the device is an initiator of the IV update procedure (look at `bt_mesh_beacon_ivu_initiator`). Test mode extension doesn't use this option just sends signals.

Reply:
I see, thanks for the detailed explanation :+1: